Source/SwiftLintBuiltInRules/Rules/Idiomatic/RedundantTypeAnnotationRule.swift

```diff
@@ -168,47 +183,69 @@ private extension RedundantTypeAnnotationRule {
+                end: typeAnnotation.endPositionBeforeTrailingTrivia,
+                replacement: ""
+            ))
         }
 
         override func visitPost(_ node: OptionalBindingConditionSyntax) {
-            if let typeAnnotation = node.typeAnnotation,
-               let initializer = node.initializer?.value,
-               typeAnnotation.isRedundant(with: initializer) {
-                violations.append(typeAnnotation.positionAfterSkippingLeadingTrivia)
-                violationCorrections.append(ViolationCorrection(
-                    start: typeAnnotation.position,
-                    end: typeAnnotation.endPositionBeforeTrailingTrivia,
-                    replacement: ""
-                ))
+            guard let typeAnnotation = node.typeAnnotation,
+                  let initializer = node.initializer?.value else {
+                return
             }
+            let isRedundant: Bool = typeAnnotation.isRedundant(
+                with: initializer,
+                considerLiteralsRedundant: configuration.considerDefaultLiteralTypesRedundant
+            )
+            guard isRedundant else {
+                return
+            }
+            violations.append(typeAnnotation.positionAfterSkippingLeadingTrivia)
+            violationCorrections.append(ViolationCorrection(
+                start: typeAnnotation.position,
+                end: typeAnnotation.endPositionBeforeTrailingTrivia,
+                replacement: ""
+            ))
         }
     }
 }
 
 private extension TypeAnnotationSyntax {
-    func isRedundant(with initializerExpr: ExprSyntax) -> Bool {
+    func isRedundant(with initializerExpr: ExprSyntax, considerLiteralsRedundant: Bool) -> Bool {
         var initializer = initializerExpr
         if let forceUnwrap = initializer.as(ForceUnwrapExprSyntax.self) {
             initializer = forceUnwrap.expression
         }
-
-        // If the initializer is a boolean expression, we consider using the `Bool` type
-        // annotation as redundant.
-        if initializer.is(BooleanLiteralExprSyntax.self) {
-            return type.trimmedDescription == "Bool"
+        // Consider the type annotation redundant if `considerLiteralsRedundant` is true and the expression is one of
+        // the supported compiler-inferred literals.
+        if considerLiteralsRedundant {
+            return switch initializer {
+            case let expr where expr.is(BooleanLiteralExprSyntax.self):
+                type.trimmedDescription == "Bool"
+            case let expr where expr.is(FloatLiteralExprSyntax.self):
+                type.trimmedDescription == "Double"
+            case let expr where expr.is(IntegerLiteralExprSyntax.self):
+                type.trimmedDescription == "Int"
+            case let expr where expr.is(StringLiteralExprSyntax.self):
+                type.trimmedDescription == "String"
+            default:
+                false
+            }
         }
```

Comment:
This part should be a separate method that's only called when the option is `true`.